File: LeetCode/513找树左下角的值.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var maxDepth int
var ans int

func calc(root *TreeNode, depth int) {
	if root.Left == nil && root.Right == nil {
		if depth > maxDepth {
			maxDepth = depth
			ans = root.Val
		}
		return
	}
	if root.Left != nil {
		depth++
		calc(root.Left, depth)
		depth--
	}
	if root.Right != nil {
		depth++
		calc(root.Right, depth)
		depth--
	}
	return
}

func findBottomLeftValue(root *TreeNode) int {
	if root.Left == nil && root.Right == nil {
		return root.Val
	}
	calc(root, maxDepth)
	return ans
}

/*
	递归三个步骤
	1、找出递归参数、递归函数返回值
	2、确定终止条件
	3、确定单层循环逻辑
*/

File: LeetCode/513找树左下角的值.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findBottomLeftValue(TreeNode root) {
        Queue<TreeNode> que = new LinkedList<>();
        if (root != null) que.add(root);
        int leftVal = 0;
        while (!que.isEmpty()) {
            int size = que.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = que.poll();
                if (i == 0) leftVal = node.val;
                if (node.left != null) que.add(node.left);
                if (node.right != null) que.add(node.right);
            }
        }
        return leftVal;
    }
}